src/gas/fetchBlockFeeHistory.ts

```diff
@@ -0,0 +1,235 @@
+ * on how this works.)
+ */
+type Block<Percentile extends number> = {
+  number: BN;
+  baseFeePerGas: BN;
+  gasUsedRatio: number;
+  priorityFeesByPercentile: Record<Percentile, BN>;
+};
+
+const MAX_NUMBER_OF_BLOCKS_PER_ETH_FEE_HISTORY_CALL = 1024;
+
+/**
+ * Uses `eth_feeHistory` (an EIP-1559 feature) to obtain information about gas fees from a range of
+ * blocks that have occurred recently on a network.
+ *
+ * To learn more, see these resources:
+ *
+ * - <https://infura.io/docs/ethereum#operation/eth_feeHistory>
+ * - <https://github.com/zsfelfoldi/feehistory/blob/main/docs/feeHistory.md>
+ * - <https://github.com/ethereum/go-ethereum/blob/57a3fab8a75eeb9c2f4fab770b73b51b9fe672c5/eth/gasprice/feehistory.go#L180>
+ * - <https://github.com/ethereum/EIPs/blob/master/EIPS/eip-1559.md>
+ * - <https://gas-api.metaswap.codefi.network/testFeeHistory>
+ *
+ * @param args - The arguments to this function.
+ * @param args.ethQuery - An EthQuery instance that wraps a provider for the network in question.
+ * @param args.endBlock - The desired end of the requested block range. Can be "latest" if you want
+ * to start from the latest successful block or the number of a known past block.
+ * @param args.numberOfBlocks - How many total blocks to fetch.
+ * @param args.percentiles - A set of numbers between 1 and 100 which will dictate how
+ * `priorityFeesByPercentile` in each returned block will be formed. When Ethereum runs the
+ * `eth_feeHistory` method, for each block it is considering, it will first sort all transactions by
+ * the priority fee. It will then go through each transaction and add the total amount of gas paid
+ * for that transaction to a bucket which maxes out at the total gas used for the whole block. As
+ * the bucket fills, it will cross percentages which correspond to the percentiles specified here,
+ * and the priority fees of the first transactions which cause it to reach those percentages will be
+ * recorded. Hence, `priorityFeesByPercentile` represents the priority fees of transactions at key
+ * gas used contribution levels, where earlier levels have smaller contributions and later levels
+ * have higher contributions.
+ * @returns The list of blocks and their fee data, sorted from oldest to newest.
+ */
+export default async function fetchBlockFeeHistory<Percentile extends number>({
+  ethQuery,
+  numberOfBlocks: totalNumberOfBlocks,
+  endBlock: givenEndBlock = 'latest',
+  percentiles: givenPercentiles = [],
+}: {
+  ethQuery: EthQuery;
+  numberOfBlocks: number;
+  endBlock?: 'latest' | BN;
+  percentiles?: readonly Percentile[];
+}): Promise<Block<Percentile>[]> {
+  const percentiles =
+    givenPercentiles.length > 0
+      ? Array.from(new Set(givenPercentiles)).sort((a, b) => a - b)
+      : [];
+
+  const finalEndBlockNumber =
+    givenEndBlock === 'latest'
+      ? await query(ethQuery, 'blockNumber')
+      : givenEndBlock;
+
+  const requestChunkSpecifiers = determineRequestChunkSpecifiers(
```

Comment:
In case the `numberOfBlocks` is more than 1024, this function will make multiple calls to `eth_feeHistory` in order to gather all of the desired blocks. This comes in handy later when we want to provide a fallback to Sam's "busy threshold" endpoint, which requires fetching 20,000 blocks.